### PR TITLE
linux/linkage_checker: remove `gcc` from `undeclared_deps`

### PR DESCRIPTION
**Pull request by carlocab**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

This is causing warnings all over Homebrew/homebrew-core.

For example, there are dozens of warnings about this in the following CI
run: https://github.com/Homebrew/homebrew-core/runs/8116257943?check_suite_focus=true

### CODE REVIEW

**Comment by MikeMcQuaid** [comment]
@carlocab @danielnachun @iMichka @sjackman I'd like to understand a little more why this is happening. Are these opportunistic linkages? Are these actually using the GCC formula but it's just added as a dependency in a different way? Will this be resolved when we rebuild bottles on 22.04?

If this is temporary until 22.04 migration: I'm fine with adding this now with a Ruby comment noting when it should be removed. If this is not temporary: can we work out how to resolve this properly?

**Review by MikeMcQuaid — APPROVED** [review]
Ok, I'm fine as-is given the explanation. Would love to see an inline Ruby comment summarising this situation and explaining why with a `TODO` that links to an issue describing the work that needs to be done. Thanks again @carlocab!

**Comment by danielnachun** [comment]
>This happens because, when foo is built with gcc, foo acquires a dependency on gcc's runtime libraries, and its dependents must then also use the same runtime libraries or else they break. https://github.com/Homebrew/brew/pull/13659 makes it so that those dependents can use gcc's runtime libraries without being rebuilt.

I think https://github.com/Homebrew/brew/pull/13659 is a great idea and we should keep it.  Although I think we should avoid using brewed GCC whenever possible, in the rare case it may be needed preventing the dependency explosion is critical.

> Step 2 above is, from my understanding, planned, and will be enabled by building bottles on 22.04. Step 3 will likely happen eventually.

I will be working as much as possible to get rid of all the Linux-only `gcc` dependencies as quickly as possible once the migration is done.  That will get rid of the problem unless we eventually have a formula that needs to be built with brewed GCC and has C++ dependents.

> An alternative to the above would be to make the gcc linkage check a bit more sophisticated by examining the objects that seem to link with gcc to check if they indeed need the latest gcc or can manage with our preferred GCC's runtime libraries.
>
> I know @sjackman had ideas about doing something like this for libstdc++ -- I'm not sure, however, if this is enough to cover every type of "optional" linkage with gcc, since there could be linkage to other runtime libraries as well (though libstdc++ will be the most common).

This is the proper solution and I will go over the example @sjackman gave on doing this to see how to implement it best. The logic essentially boils down to finding the newest symbol ABI version in the bottle's binaries and making sure it is not newer than what is provided by GCC 11.  It only applies to 4 libraries `libstdc++`, `libgomp`, `libgcc_s` and `libatomic`, and for a given bottle we should only check the libraries among those 4 that are actually linked.

File: Library/Homebrew/extend/os/linux/linkage_checker.rb [hunk 1]
```diff
@@ -80,5 +80,14 @@ def check_dylibs(rebuild_cache:)
...
       SYSTEM_LIBRARY_ALLOWLIST.include? File.basename(s)
     end
+
+    # We build all formulae with an RPATH that includes the gcc formula's runtime lib directory.
+    # See: https://github.com/Homebrew/brew/blob/e689cc07/Library/Homebrew/extend/os/linux/extend/ENV/super.rb#L53
+    # This results in formulae showing linkage with gcc whenever it is installed, even if no dependency is declared.
+    # See discussions at:
+    #   https://github.com/Homebrew/brew/pull/13659
+    #   https://github.com/Homebrew/brew/pull/13796
+    # TODO: Find a nicer way to handle this. (e.g. examining the ELF file to determine the required libstdc++.)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/linkage_checker.rb:90]: @danielnachun can you tackle this post 3.6.0? Please open a tracking issue if that'd be helpful for you to remember 👍🏻

Thanks ❤️

**Pull request merged**